eth_account/_utils/encode_typed_data/encoding_and_hashing.py

```diff
@@ -0,0 +1,209 @@
+from typing import (
+    Any,
+    Dict,
+    List,
+    Tuple,
+    Union,
+)
+
+from eth_abi import (
+    encode,
+)
+from eth_utils import (
+    is_hexstr,
+    keccak,
+    to_bytes,
+    to_int,
+)
+
+from eth_account._utils.encode_typed_data.helpers import (
+    EIP712_SOLIDITY_TYPES,
+    coerce_bool,
+    is_array_type,
+    parse_core_array_type,
+    parse_parent_array_type,
+)
+
+
+def get_primary_type(types: Dict[str, List[Dict[str, str]]]) -> str:
+    custom_types = set(types.keys())
+    custom_types_that_are_deps = set()
+
+    for type_ in custom_types:
+        type_fields = types[type_]
+        for field in type_fields:
+            parsed_type = parse_core_array_type(field["type"])
+            if parsed_type in custom_types and parsed_type != type_:
+                custom_types_that_are_deps.add(parsed_type)
+
+    primary_type = list(custom_types.difference(custom_types_that_are_deps))
+    if len(primary_type) == 1:
+        return primary_type[0]
+    else:
+        raise ValueError("Unable to determine primary type")
+
+
+def encode_field(
+    types: Dict[str, List[Dict[str, str]]],
+    name: str,
+    type_: str,
+    value: Any,
+) -> Tuple[str, Union[int, bytes]]:
+    if type_ in types.keys():
+        # type is a custom type
+        if value is None:
+            return ("bytes32", b"\x00" * 32)
+        else:
+            return ("bytes32", keccak(encode_data(type_, types, value)))
+
+    elif type_ in ["string", "bytes"] and value is None:
+        return ("bytes32", b"")
+
+    # None is allowed only for custom and dynamic types
+    elif value is None:
+        raise ValueError(f"Missing value for field `{name}` of type `{type_}`")
+
+    elif is_array_type(type_):
+        type_ = parse_parent_array_type(type_)
+        type_value_pairs = [encode_field(types, name, type_, item) for item in value]
+        data_types, data_hashes = zip(*type_value_pairs)
+        return ("bytes32", keccak(encode(data_types, data_hashes)))
+
+    elif type_ == "bool":
+        return (type_, coerce_bool(value))
+
+    elif type_ == "bytes":
+        if is_hexstr(value):
+            value = to_bytes(hexstr=value)
+        else:
+            value = to_bytes(value)
+        return ("bytes32", keccak(value))
+
+    elif type_ == "string":
+        if isinstance(value, int):
+            value = to_bytes(value)
+        else:
+            value = to_bytes(text=value)
+        return ("bytes32", keccak(value))
+
+    # allow string values for int and uint types
+    elif type(value) == str and type_.startswith(("int", "uint")):
+        if value[:2] == "0x":
+            return (type_, to_int(hexstr=value))
+        else:
+            return (type_, to_int(text=value))
+
```

Comment:
I'm not sure on this one tbh... but if we want to be a bit more forgiving with `address` type as well, we could add something like:

```py
    elif type_ == "address":
        if is_0x_prefixed_hexstr(value):
            value = to_bytes(hexstr=value)
        elif isinstance(value, int):
            value = to_bytes(value)

        if len(value) < 20:
            # pad values that are not 20 bytes
            value = value.rjust(20, b"\x00")

        return (type_, value)
```

This would pass some of the `eth-sig-util` test cases such as encoding type `address` with value:

- `0x0`
- `0x1fffffffffffff`
- `0x10`
- `10`

etc... which `eth-sig-util` tests against

Reply:
I feel like address should be less forgiving in this regard. The 0x-prefixed string vs non-prefixed I can get behind, but it really seems like if a message has come through with an address of "0x10" that a mistake has been made somewhere.

Reply:
I agree with you there though. I'm OK not adding any of this, even non-0x prefixed. It seems to give different results in `eth-sig-util`. We can leave this out entirely, just figured I'd put it here for discussion.